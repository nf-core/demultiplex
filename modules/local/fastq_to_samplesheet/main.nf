process FASTQ_TO_SAMPLESHEET {
    tag "$meta.id"

    executor 'local'
    memory 100.MB

    input:
    val meta // Expecting a list of items
    val pipeline
    val strandedness

    output:
    tuple val(meta_clone), path("*samplesheet.csv"), emit: samplesheet

    exec:
    // Initialize the samplesheet content
    def samplesheetHeader = []
    def samplesheetRows = []

    // Sort meta by item.id
    def sortedMeta = meta.sort { it.id }

    // Collect all unique columns from all items and create rows
    def allColumns = new LinkedHashSet()

    sortedMeta.each { item ->
        // Check for required keys in each item
        if (!item.samplename) {
            error "Item with id ${item.id} is missing the 'samplename' key."
        }
        if (!item.fastq_1) {
            error "Item with id ${item.id} is missing the 'fastq_1' key."
        }

        def pipeline_map = [:] // Initialize as an empty map

        // Prepare sample information
        pipeline_map.sample = item.samplename
        pipeline_map.fastq_1 = item.publish_dir + '/' + file(item.fastq_1).fileName

        // Add fastq_2 if it's a paired-end sample
        if (!item.single_end && item.fastq_2) {
            pipeline_map.fastq_2 = item.publish_dir + '/' + file(item.fastq_2).fileName ?: ''
        }

        // Add pipeline-specific entries
        if (pipeline == 'rnaseq') {
            pipeline_map.strandedness = strandedness ?: ''
        } else if (pipeline == 'atacseq') {
            pipeline_map.replicate = 1
        } else if (pipeline == 'taxprofiler') {
            pipeline_map.fasta = ''
        }

        // Add all keys to the set of unique columns
        allColumns.addAll(pipeline_map.keySet())

        // Prepare a row for the samplesheet, filling in missing values with empty strings
        def rowValues = allColumns.collect { key ->
            pipeline_map.containsKey(key) ? '"' + pipeline_map[key] + '"' : '""'
        }
        samplesheetRows << rowValues.join(",")
    }

    // Create a sorted list of headers
    samplesheetHeader = allColumns.collect { '"' + it + '"' }

    // Create the complete samplesheet content
    def samplesheet = samplesheetHeader.join(",") + '\n' + samplesheetRows.join("\n")

    // Write samplesheet to file
    def samplesheet_file = task.workDir.resolve("${pipeline}_samplesheet.csv")
    samplesheet_file.text = samplesheet

    // Clone the first item in meta for output
    meta_clone = meta.first().clone()
    meta_clone.remove('publish_dir') // Removing the publish_dir just in case, although output channel is not used by other process

}
