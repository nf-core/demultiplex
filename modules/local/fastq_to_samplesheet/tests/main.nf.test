nextflow_process {

    name "Test Process FASTQ_TO_SAMPLESHEET"
    script "../main.nf"
    process "FASTQ_TO_SAMPLESHEET"

    tag "modules"
    tag "modules_local"
    tag "fastq_to_samplesheet"

    test("Should run without failures") {

        when {
            process {
                """
                input[0] = Channel.of([[id:'Sample1_S1_L001', samplename:'Sample1', fcid:'220422_M11111_0222_000000000-K9H97', lane:'1', empty:false, single_end:true, fastq_1:'Sample1_S1_L001_R1_001.fastq.gz']])
                input[1] = 'rnaseq'
                input[2] = 'auto'
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
