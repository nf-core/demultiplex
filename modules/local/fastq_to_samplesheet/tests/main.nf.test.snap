{
    "Should run without failures": {
        "content": [
            {
                "0": [
                    [
                        [
                            {
                                "id": "Sample1_S1_L001",
                                "samplename": "Sample1",
                                "fcid": "220422_M11111_0222_000000000-K9H97",
                                "lane": "1",
                                "empty": false,
                                "single_end": true,
                                "fastq_1": "Sample1_S1_L001_R1_001.fastq.gz"
                            }
                        ],
                        "[Sample1_S1_L001].samplesheet.csv:md5,bc779a8b2302a093cbb04a118bb5c90f"
                    ]
                ],
                "samplesheet": [
                    [
                        [
                            {
                                "id": "Sample1_S1_L001",
                                "samplename": "Sample1",
                                "fcid": "220422_M11111_0222_000000000-K9H97",
                                "lane": "1",
                                "empty": false,
                                "single_end": true,
                                "fastq_1": "Sample1_S1_L001_R1_001.fastq.gz"
                            }
                        ],
                        "[Sample1_S1_L001].samplesheet.csv:md5,bc779a8b2302a093cbb04a118bb5c90f"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-08-09T22:00:18.282617632"
    }
}